AsyncDisplayKit/Layout/ASLayoutable.mm

```diff
@@ -38,3 +45,183 @@ struct ASLayoutableContext ASLayoutableGetCurrentContext()
+ *
+ *  While this could be done in the layoutable object itself, moving the logic into this helper function
+ *  allows a custom spec to set up defaults without needing to alter the layoutable itself. For example,
+ *  image you were creating a custom baseline spec that needed ascender/descender. To assign values automatically
+ *  when a text node's attribute string is set, you would need to subclass ASTextNode and assign the values in the
+ *  override of setAttributeString. However, assigning the defaults via this function allows you to create a
+ *  custom spec without the need to create a subclass of ASTextNode.
+ *
+ *  @param layoutable The layoutable object to inspect for default intrinsic layout option values
+ */
+void ASLayoutableSetValuesForLayoutable(id<ASLayoutable> layoutable)
+{
+  //ASDN::MutexLocker l(_propertyLock);
+  if ([layoutable isKindOfClass:[ASDisplayNode class]]) {
+    ASDisplayNode *displayNode = (ASDisplayNode *)layoutable;
+    displayNode.sizeRange = ASRelativeSizeRangeMake(ASRelativeSizeMakeWithCGSize(displayNode.preferredFrameSize), ASRelativeSizeMakeWithCGSize(displayNode.preferredFrameSize));
+    
+    if ([layoutable isKindOfClass:[ASTextNode class]]) {
+      ASTextNode *textNode = (ASTextNode *)layoutable;
+      NSAttributedString *attributedString = textNode.attributedString;
+      if (attributedString.length > 0) {
+        CGFloat screenScale = ASScreenScale();
+        textNode.ascender = round([[attributedString attribute:NSFontAttributeName atIndex:0 effectiveRange:NULL] ascender] * screenScale)/screenScale;
+        textNode.descender = round([[attributedString attribute:NSFontAttributeName atIndex:attributedString.length - 1 effectiveRange:NULL] descender] * screenScale)/screenScale;
+      }
+    }
+    
+  }
+}
+
+
+#pragma mark - ASLayoutOptionsForwarding
+
+/**
+ *  Both an ASDisplayNode and an ASLayoutSpec conform to ASLayoutable. There are several properties
+ *  in ASLayoutable that are used when a node or spec is used in a layout spec.
+ *  These properties are provided for convenience, as they are forwards to the node or spec's
+ *  properties. Instead of duplicating the property forwarding in both classes, we
+ *  create a define that allows us to easily implement the forwards in one place.
+ *
+ *  If you create a custom layout spec, we recommend this stragety if you decide to extend
+ *  ASDisplayNode and ASLayoutSpec to provide convenience properties for any options that your
+ *  layoutSpec may require.
+ */
+
+#define ASEnvironmentLayoutOptionsForwarding \
+- (ASEnvironmentLayoutOptionsState *)layoutOptionsState\
+{\
+  return &(self.environmentCollection->layoutOptionsState);\
+}\
+- (void)propagateUpLayoutOptionsState\
+{\
+  id<ASEnvironment> parent = [self parent];\
+  if (![parent supportsMultipleChildren]) {\
+    ASEnvironmentStatePropagateUp(parent, self.environmentCollection->layoutOptionsState);\
+  }\
+}\
+\
+- (CGFloat)spacingAfter\
```

Comment:
All of this properties needs to be locked with the property locks of the classes:

Setters: instance variable .lock() and .unlock()
Getter: MutexLockers